File: src/assertions.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const types = @import("types.zig");
const Assertion = types.Assertion;
const AssertionResult = types.AssertionResult;
const HttpResult = types.HttpResult;

pub fn evaluateAssertions(allocator: Allocator, assertions: []const Assertion, result: *const HttpResult) !std.ArrayList(AssertionResult) {
    var assertion_results = std.ArrayList(AssertionResult).init(allocator);

    for (assertions) |assertion| {
        const assertion_result = try evaluateAssertion(allocator, assertion, result);
        try assertion_results.append(assertion_result);
    }

    return assertion_results;
}

fn evaluateAssertion(allocator: Allocator, assertion: Assertion, result: *const HttpResult) !AssertionResult {
    switch (assertion.type) {
        .response_status => {
            const expected_status = std.fmt.parseInt(u16, assertion.expected_value, 10) catch {
                return AssertionResult{
                    .assertion = assertion,
                    .passed = false,
                    .actual_value = try std.fmt.allocPrint(allocator, "{}", .{result.status_code}),
                    .error_message = try allocator.dupe(u8, "Invalid expected status code format"),
                };
            };

            const passed = result.status_code == expected_status;
            return AssertionResult{
                .assertion = assertion,
                .passed = passed,
                .actual_value = try std.fmt.allocPrint(allocator, "{}", .{result.status_code}),
                .error_message = if (!passed)
                    try std.fmt.allocPrint(allocator, "Expected status {}, got {}", .{ expected_status, result.status_code })
                else
                    null,
            };
        },

        .response_body => {
            if (result.response_body) |body| {
                const passed = std.mem.indexOf(u8, body, assertion.expected_value) != null;
                return AssertionResult{
                    .assertion = assertion,
                    .passed = passed,
                    .actual_value = try allocator.dupe(u8, body),
                    .error_message = if (!passed)
                        try std.fmt.allocPrint(allocator, "Expected body to contain '{s}'", .{assertion.expected_value})
                    else
                        null,
                };
            } else {
                return AssertionResult{
                    .assertion = assertion,
                    .passed = false,
                    .actual_value = try allocator.dupe(u8, ""),
                    .error_message = try allocator.dupe(u8, "No response body available"),
                };
            }
        },

        .response_headers => {
            if (result.response_headers) |headers| {
                const colon_pos = std.mem.indexOf(u8, assertion.expected_value, ":") orelse {
                    return AssertionResult{
                        .assertion = assertion,
                        .passed = false,
                        .actual_value = try formatHeaders(allocator, headers),
                        .error_message = try allocator.dupe(u8, "Invalid header format, expected 'Name: Value'"),
                    };
                };

                const expected_name = std.mem.trim(u8, assertion.expected_value[0..colon_pos], " \t");
                const expected_value = std.mem.trim(u8, assertion.expected_value[colon_pos + 1 ..], " \t");

                var found = false;
                for (headers) |header| {
                    if (std.mem.eql(u8, header.name, expected_name)) {
                        if (std.mem.indexOf(u8, header.value, expected_value) != null) {
                            found = true;
                            break;
                        }
                    }
                }

                return AssertionResult{
                    .assertion = assertion,
                    .passed = found,
                    .actual_value = try formatHeaders(allocator, headers),
                    .error_message = if (!found)
                        try std.fmt.allocPrint(allocator, "Expected header '{s}' with value containing '{s}'", .{ expected_name, expected_value })
                    else
                        null,
                };
            } else {
                return AssertionResult{
                    .assertion = assertion,
                    .passed = false,
                    .actual_value = try allocator.dupe(u8, ""),
                    .error_message = try allocator.dupe(u8, "No response headers available"),
                };
            }
        },
    }
}

fn formatHeaders(allocator: Allocator, headers: []const HttpResult.Header) ![]u8 {
    var formatted = std.ArrayList(u8).init(allocator);
    defer formatted.deinit();

    for (headers, 0..) |header, i| {
        if (i > 0) {
            try formatted.appendSlice(", ");
        }
        try formatted.appendSlice(header.name);
        try formatted.appendSlice(": ");
        try formatted.appendSlice(header.value);
    }

    return try allocator.dupe(u8, formatted.items);
}

File: src/assertions.rs
use crate::types::{Assertion, AssertionResult, AssertionType, HttpResult};

pub fn evaluate_assertions(assertions: &[Assertion], result: &HttpResult) -> Vec<AssertionResult> {
    assertions
        .iter()
        .map(|assertion| evaluate_assertion(assertion, result))
        .collect()
}

fn evaluate_assertion(assertion: &Assertion, result: &HttpResult) -> AssertionResult {
    match assertion.assertion_type {
        AssertionType::ResponseStatus => {
            let expected_status = match assertion.expected_value.parse::<u16>() {
                Ok(status) => status,
                Err(_) => {
                    return AssertionResult {
                        assertion: assertion.clone(),
                        passed: false,
                        actual_value: Some(result.status_code.to_string()),
                        error_message: Some("Invalid expected status code format".to_string()),
                    };
                }
            };

            let passed = result.status_code == expected_status;
            AssertionResult {
                assertion: assertion.clone(),
                passed,
                actual_value: Some(result.status_code.to_string()),
                error_message: if !passed {
                    Some(format!(
                        "Expected status {}, got {}",
                        expected_status, result.status_code
                    ))
                } else {
                    None
                },
            }
        }

        AssertionType::ResponseBody => {
            if let Some(ref body) = result.response_body {
                let passed = body.contains(&assertion.expected_value);
                AssertionResult {
                    assertion: assertion.clone(),
                    passed,
                    actual_value: Some(body.clone()),
                    error_message: if !passed {
                        Some(format!(
                            "Expected body to contain '{}'",
                            assertion.expected_value
                        ))
                    } else {
                        None
                    },
                }
            } else {
                AssertionResult {
                    assertion: assertion.clone(),
                    passed: false,
                    actual_value: Some(String::new()),
                    error_message: Some("No response body available".to_string()),
                }
            }
        }

        AssertionType::ResponseHeaders => {
            if let Some(ref headers) = result.response_headers {
                let colon_pos = assertion.expected_value.find(':');
                if colon_pos.is_none() {
                    return AssertionResult {
                        assertion: assertion.clone(),
                        passed: false,
                        actual_value: Some(format_headers(headers)),
                        error_message: Some(
                            "Invalid header format, expected 'Name: Value'".to_string(),
                        ),
                    };
                }

                let colon_pos = colon_pos.unwrap();
                let expected_name = assertion.expected_value[..colon_pos].trim();
                let expected_value = assertion.expected_value[colon_pos + 1..].trim();

                let mut found = false;
                for (name, value) in headers {
                    if name.eq_ignore_ascii_case(expected_name) {
                        if value.contains(expected_value) {
                            found = true;
                            break;
                        }
                    }
                }

                AssertionResult {
                    assertion: assertion.clone(),
                    passed: found,
                    actual_value: Some(format_headers(headers)),
                    error_message: if !found {
                        Some(format!(
                            "Expected header '{}' with value containing '{}'",
                            expected_name, expected_value
                        ))
                    } else {
                        None
                    },
                }
            } else {
                AssertionResult {
                    assertion: assertion.clone(),
                    passed: false,
                    actual_value: Some(String::new()),
                    error_message: Some("No response headers available".to_string()),
                }
            }
        }
    }
}

fn format_headers(headers: &std::collections::HashMap<String, String>) -> String {
    headers
        .iter()
        .map(|(k, v)| format!("{}: {}", k, v))
        .collect::<Vec<_>>()
        .join(", ")
}
